device/src/region/mod.rs

```diff
@@ -175,6 +199,101 @@ impl Configuration {
         Configuration::with_state(State::new(region))
     }
 
+    /// Create a new [`Configuration`] with a specific set of channels enabled
+    /// for joining the network. You can specify up to 16 preferred channels.
+    ///
+    /// When `join` is called on a [`Configuration`] created using this
+    /// method, the network will be attempted to be joined only on the provided
+    /// channel subset. This set of channels will be retried the number of times
+    /// specified; after which we will revert to trying to join with all
+    /// channels enabled using a preset sequence.
+    ///
+    /// This method only makes sense for fixed channel plans (AU915, US915).
+    /// Trying to call this constructor with a dynamic channel region will
+    /// return `Err`.
+    ///
+    /// # About supported channels (fixed channel plans only)
+    ///
+    /// Supported channels:
+    ///
+    /// * 64 125 kHz channels (0-63)
+    /// * 8 500 kHz channels (64-71)
+    ///
+    /// If a channel out of this range is specified, `Err(())` will be returned.
+    ///
+    /// # ⚠️Warning⚠️
+    ///
+    /// It is recommended to set a low number (ie, < 10) of join retries using the
+    /// preferred channels. The reason for this is if you *only* try to join
+    /// with a channel bias, and the network is configured to use a
+    /// strictly different set of channels than the ones you provide, the
+    /// network will NEVER be joined.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(Configuration)` if the provided channel set is correct and the
+    ///   region is a fixed channel plan
+    /// * The length of `channel_list` must be <= 16, otherwise `Err` will
+    ///   be returned.
+    /// * If a channel out of the specified channel range is specified,
+    ///   `Err` will be returned (ie, >= 72).
+    pub fn new_with_preferred_channels(
+        region: Region,
+        preferred_channels: &[u8],
+        num_retries: usize,
+    ) -> Result<Configuration, Error> {
+        use Region::*;
+        match region {
+            US915 | AU915 => {
```

Comment:
The existence of this functions seems like a code-smell to me.

Now that I think about it, this logic might be scoped wrong and might belong in fixed_channel_plans/mod.rs instead? If that works, then you can avoid the `UnsupportedRegion` as well.

Reply:
The problem is that whether the actual function lives in `region/mod.rs` or `fixed_channel_plans/mod.rs`, at the lowest level the end user can only construct a `Region`. That enum makes no distinction between fixed and dynamic channel plans. There would probably be a way to type-encode that, but it would probably involve creating more intermediary types, and I'm not sure the added complexity is so worth it.

Reply:
I think we could look at the [`with_state`](https://github.com/ivajloip/rust-lorawan/blob/master/device/src/region/mod.rs#L178) function. Either use it directly or do something in that theme.

Essentially, the user could construct US915/AU915 "state" and do the join bias and then build a configuration there. Really, we should probably just implement From<Region> for Configuration for all "states", then the user can make tweaks and "into" it when constructing a device.

Reply:
Sorry if my last thought there is not clear - I have limited time today. I have a flight tmrw and may try to draft what I mean and PR to your branch.

Generally though, I think it’s quite possible to make these errors compile-time impossible without much complexity. I think that’s always better design than throwing errors are runtime.

Reply:
Threw up this PR to show what I mean: https://github.com/jbeaurivage/rust-lorawan/pull/1

Reply:
@lthiery, take a look at the latest revision. I've managed to make specifying preferred channels completely statically checked. Although, now the user has to provide a reference to a fixed-size array rather than a slice. I think the benefits outweigh the hit on ergonomics, but I'd definitely like your opinion on the matter.